modules/lang-painless/src/main/java/org/elasticsearch/painless/DecorateExecutePhase.java

```diff
@@ -0,0 +1,310 @@
+import org.elasticsearch.painless.ir.ThrowNode;
+import org.elasticsearch.painless.ir.TryNode;
+import org.elasticsearch.painless.ir.UnboundCallNode;
+import org.elasticsearch.painless.ir.UnboundFieldNode;
+import org.elasticsearch.painless.ir.VariableNode;
+import org.elasticsearch.painless.lookup.PainlessLookup;
+import org.elasticsearch.painless.lookup.PainlessMethod;
+import org.elasticsearch.painless.symbol.FunctionTable.LocalFunction;
+import org.elasticsearch.painless.symbol.ScriptRoot;
+import org.elasticsearch.script.ScriptException;
+import org.objectweb.asm.commons.Method;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * This injects additional ir nodes required to for
+ * the "execute" method. This includes injection of ir nodes
+ * to convert get methods into local variables for those
+ * that are used and adds additional sandboxing by wrapping
+ * the main "execute" block with several exceptions.
+ */
+public class DecorateExecutePhase {
+
+    public static void phase(ScriptRoot scriptRoot, ClassNode classNode) {
+        FunctionNode executeFunctionNode = null;
+
+        // look up the execute method for decoration
+        for (FunctionNode functionNode : classNode.getFunctionsNodes()) {
+            if ("execute".equals(functionNode.getName())) {
+                executeFunctionNode = functionNode;
+                break;
+            }
+        }
+
+        BlockNode blockNode = executeFunctionNode.getBlockNode();
+
+        // convert gets methods to a new set of inserted ir nodes as necessary -
+        // requires the gets method name be modified from "getExample" to "example"
+        // if a get method variable isn't used it's declaration node is removed from
+        // the ir tree permanently so there is no frivolous variable slotting
+        int statementIndex = 0;
+
+        while (statementIndex < blockNode.getStatementsNodes().size()) {
+            StatementNode statementNode = blockNode.getStatementsNodes().get(statementIndex);
+
+            if (statementNode instanceof DeclarationNode) {
+                DeclarationNode declarationNode = (DeclarationNode)statementNode;
+                boolean isRemoved = false;
+
+                for (int getIndex = 0; getIndex < scriptRoot.getScriptClassInfo().getGetMethods().size(); ++getIndex) {
+                    Class<?> returnType = scriptRoot.getScriptClassInfo().getGetReturns().get(getIndex);
+                    Method getMethod = scriptRoot.getScriptClassInfo().getGetMethods().get(getIndex);
+                    String name = getMethod.getName().substring(3);
+                    name = Character.toLowerCase(name.charAt(0)) + name.substring(1);
+
+                    if (name.equals(declarationNode.getName())) {
+                        if (scriptRoot.getUsedVariables().contains(name)) {
+                                UnboundCallNode unboundCallNode = new UnboundCallNode();
```

Comment:
Maybe we should rename this to `MemberCallNode`? The meaning of unbound is unclear

Reply:
Yes, and fixed.